{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e5253037",
   "metadata": {},
   "source": [
    "# Quantization of difference between 3D protein structures\n",
    "\n",
    "In this project, we focus on quantitative representation of difference between a pair of protein structures. \n",
    "\n",
    "### Prerequisites\n",
    "\n",
    "1. Mapping from isoforms(RNA) to proteins(Sequence) is a bijection.  \n",
    "\n",
    "2. Mapping from gene to isoform is one-to-many. \n",
    "\n",
    "3. Any two pair of isoforms transcribed by the same genome will have overlaps among sequences of their proteins. \n",
    "\n",
    "### What we need\n",
    "\n",
    "1. A map from gene -> isoforms -> proteins -> atoms \n",
    "\n",
    "2. A map from gene -> exon -> length of the exon (is this number of atoms?)\n",
    "\n",
    "### Algorithm for technique 1\n",
    "\n",
    "1. Given sequencing of two proteins (say P<sub>1</sub> and P<sub>2</sub>) which have originated from a single genome, say G.\n",
    "\n",
    "2. For each protein, say P<sup>(1)</sup>, we have a sequence of amino acids a<sup>(1)</sup><sub>1</sub>, a<sup>(1)</sup><sub>2</sub>, .... a<sup>(1)</sup><sub>n</sub> where n is the length of the protein.\n",
    "\n",
    "3. We predict structures of the proteins using alphafold2 such that P<sup>(1)</sup><sub>i</sub> are the co-ordinates of a<sup>(1)</sup><sub>i</sub>. \n",
    "\n",
    "4. Let us say that proteins P<sup>(1)</sup> and P<sup>(2)</sup> correspond to isoforms I<sup>(1)</sup> and I<sup>(2)</sup>. (Note that : I<sup>(1)</sup> and I<sup>(2)</sup> must be transcribed by the source genome G). \n",
    "\n",
    "5. We find the exons shared by I<sup>(1)</sup> and I<sup>(2)</sup>. Then by using the lengths of the exons i.e. number of amino acids in each exon, we find overlapping runs between P<sup>(1)</sup> and P<sup>(2)</sup>. \n",
    "\n",
    "6. Thus, we get k overlaps L<sub>1</sub>, L<sub>2</sub>, .... L<sub>k</sub> where each lap L<sub>i</sub> stores starting and ending indices of the overlaps in each protein. I.e. L<sub>i</sub> is represented as tuple of size 4 (i<sub>1</sub>, i<sub>2</sub>, j<sub>1</sub>, j<sub>2</sub>) such that i<sub>2</sub> - i<sub>1</sub> == j<sub>2</sub> - i<sub>1</sub> and a<sup>(1)</sup><sub>i<sub>1</sub> + x </sub> == a<sup>(2)</sup><sub>j<sub>1</sub> + x </sub> for x = 0, 1, ... j<sub>2</sub> - j<sub>1</sub>. \n",
    "\n",
    "8. Given an overlapping run for a protein, say a<sup>(1)</sup><sub>i<sub>1</sub></sub> to a<sup>(1)</sup><sub>i<sub>2</sub></sub>, we find distance between each pair of the amino acids to form a symmetric matrix D<sup>(1)</sup><sub>1</sub> such that D<sup>(1)</sup><sub>1</sub>[i][j] = ||a<sup>(1)</sup><sub>i</sub> - a<sup>(1)</sup><sub>j</sub>|| for i, j in {i<sub>1</sub>, .... i<sub>2</sub>}. \n",
    "\n",
    "9. Given all such matrices, say D<sup>(1)</sup><sub>1</sub>, D<sup>(1)</sup><sub>2</sub>, ...... D<sup>(1)</sup><sub>m</sub> for P<sub>1</sub> and D<sup>(2)</sup><sub>1</sub>, D<sup>(2)</sup><sub>2</sub>, ...... D<sup>(2)</sup><sub>m</sub> for P<sub>2</sub>, we find sum of norms of differences i.e. ||D<sup>(1)</sup><sub>i</sub> - D<sup>(2)</sup><sub>i</sub>||<sub>F</sub>. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5638fbc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import pairwise2, SeqIO\n",
    "from Bio.PDB import PDBParser\n",
    "from itertools import *\n",
    "import json\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from termcolor import colored\n",
    "from utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b754b34e",
   "metadata": {},
   "outputs": [],
   "source": [
    "root = '/Users/aasavarikakne/Desktop/cme291/'\n",
    "\n",
    "protein1_path = root + 'prediction_PKM1/'\n",
    "protein2_path = root + 'prediction_PKM2/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9d2d5c05",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence 1 has 531 amino acids.\n",
      "Sequence 2 has 531 amino acids.\n"
     ]
    }
   ],
   "source": [
    "# step 1 : read protein sequences from local storage\n",
    "seq1 = str(SeqIO.read(protein1_path + 'PKM1.fasta', \"fasta\").seq)\n",
    "seq2 = str(SeqIO.read(protein2_path + 'PKM2.fasta', \"fasta\").seq)\n",
    "\n",
    "for i, seq in enumerate([seq1, seq2]):\n",
    "    print(\"Sequence {} has {} amino acids.\".format(i+1, len(seq)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "30108b73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment starts at 0 and ends at 546.\n"
     ]
    }
   ],
   "source": [
    "# step 2 : get overlap using biopython\n",
    "alignment = max(pairwise2.align.globalxx(seq1, seq2))\n",
    "\n",
    "print(\"Alignment starts at {} and ends at {}.\".format(alignment.start, alignment.end))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8d8df1e0",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34m0 : MSKPHSEAGTAFIQTQQLHAAMADTFLEHMCRLDIDSPPITARNTGIICT\u001b[0m\n",
      "\u001b[36m0 : MSKPHSEAGTAFIQTQQLHAAMADTFLEHMCRLDIDSPPITARNTGIICT\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m1 : IGPASRSVETLKEMIKSGMNVARLNFSHGTHEYHAETIKNVRTATESFAS\u001b[0m\n",
      "\u001b[36m1 : IGPASRSVETLKEMIKSGMNVARLNFSHGTHEYHAETIKNVRTATESFAS\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m2 : DPILYRPVAVALDTKGPEIRTGLIKGSGTAEVELKKGATLKITLDNAYME\u001b[0m\n",
      "\u001b[36m2 : DPILYRPVAVALDTKGPEIRTGLIKGSGTAEVELKKGATLKITLDNAYME\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m3 : KCDENILWLDYKNICKVVEVGSKIYVDDGLISLQVKQKGADFLVTEVENG\u001b[0m\n",
      "\u001b[36m3 : KCDENILWLDYKNICKVVEVGSKIYVDDGLISLQVKQKGADFLVTEVENG\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m4 : GSLGSKKGVNLPGAAVDLPAVSEKDIQDLKFGVEQDVDMVFASFIRKASD\u001b[0m\n",
      "\u001b[36m4 : GSLGSKKGVNLPGAAVDLPAVSEKDIQDLKFGVEQDVDMVFASFIRKASD\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m5 : VHEVRKVLGEKGKNIKIISKIENHEGVRRFDEILEASDGIMVARGDLGIE\u001b[0m\n",
      "\u001b[36m5 : VHEVRKVLGEKGKNIKIISKIENHEGVRRFDEILEASDGIMVARGDLGIE\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m6 : IPAEKVFLAQKMMIGRCNRAGKPVICATQMLESMIKKPRPTRAEGSDVAN\u001b[0m\n",
      "\u001b[36m6 : IPAEKVFLAQKMMIGRCNRAGKPVICATQMLESMIKKPRPTRAEGSDVAN\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m7 : AVLDGADCIMLSGETAKGDYPLEAVRMQHLIAREAEAAMFHRKLFEELVR\u001b[0m\n",
      "\u001b[36m7 : AVLDGADCIMLSGETAKGDYPLEAVRMQHLIAREAEAAIYHLQLFEELRR\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m8 : -ASSHS--T-DLM--EAM-AM-GS-VEASY-KCLAA---ALI-VLTE-SG\u001b[0m\n",
      "\u001b[36m8 : LA----PITSD--PTEA-TA-VG-AVEAS-FKC---CSGA-IIVLT-KSG\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m9 : RSAHQVARYRPRAPIIAVTRNPQTARQAHLYRGIFPVLCKDPVQEAWAED\u001b[0m\n",
      "\u001b[36m9 : RSAHQVARYRPRAPIIAVTRNPQTARQAHLYRGIFPVLCKDPVQEAWAED\u001b[0m\n",
      "--------------------------------------------------------------------------------\n",
      "\u001b[34m10 : VDLRVNFAMNVGKARGFFKKGDVVIVLTGWRPGSGFTNTMRVVPVP\u001b[0m\n",
      "\u001b[36m10 : VDLRVNFAMNVGKARGFFKKGDVVIVLTGWRPGSGFTNTMRVVPVP\u001b[0m\n",
      "--------------------------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# debugging\n",
    "x1 = alignment.seqA\n",
    "x2 = alignment.seqB\n",
    "s = alignment.start\n",
    "e = alignment.end\n",
    "\n",
    "for i in range(s, e, 50):\n",
    "    print(colored(\"{} : {}\".format(i//50, x1[i:i+50]), 'blue'))\n",
    "    print(colored(\"{} : {}\".format(i//50, x2[i:i+50]), 'cyan'))\n",
    "    print('-'*80)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b7f1a619",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# step 3 : map alignment indices to sequence indices\n",
    "map1 = get_index_map(alignment.seqA)\n",
    "map2 = get_index_map(alignment.seqB)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8e5f8e0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# step 4 : group the aligned residues together \n",
    "overlaps = get_overlaps(alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "289226ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "for s, e in overlaps:\n",
    "    a = seq1[map1[s] : map1[e] + 1]\n",
    "    b = seq2[map2[s] : map2[e] + 1]\n",
    "    length = e - s + 1\n",
    "    for j in range(length):\n",
    "        if a[j] != b[j]:\n",
    "            print(s + j)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "31c45dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# step 5 : read output from alphafold, get positioning for each atom inside each residue\n",
    "parser = PDBParser(PERMISSIVE = True, QUIET = True) \n",
    "\n",
    "data1 = parser.get_structure(\"2fat\", protein1_path + \"selected_prediction.pdb\")\n",
    "data2 = parser.get_structure(\"2fat\", protein2_path + \"selected_prediction.pdb\")\n",
    "\n",
    "residues1 = get_residues(data1)\n",
    "residues2 = get_residues(data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "135e74b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_positions_for_overlap(residues, start, end):\n",
    "    positions = []\n",
    "    for i in range(start, end + 1):\n",
    "        for atom in residues[i].get_atoms():\n",
    "            p = atom.get_vector().get_array()\n",
    "            positions.append(p) \n",
    "    positions = np.stack(positions, axis=0) if positions else None\n",
    "    return positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "37cee671",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# step 6 : for each overlap, get the positioning all atoms in the overlap for each protein\n",
    "y = []\n",
    "x = []\n",
    "\n",
    "for s, e in overlaps:\n",
    "    # m amino acid \n",
    "    x.append(e - s + 1)\n",
    "    \n",
    "    p1 = get_position_for_overlap(residues1, map1[s], map1[e]) # n x 3\n",
    "    p2 = get_position_for_overlap(residues2, map2[s], map2[e]) # n x 3\n",
    "\n",
    "    D1 = get_dist_mat(p1) # D1[i, j] = || p1[i] - p1[j] || \n",
    "    D2 = get_dist_mat(p2)\n",
    "    n = p1.shape[0]\n",
    "\n",
    "    y.append(np.linalg.norm(D1 - D2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "be7a96a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7fa5434faaf0>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYEAAAD4CAYAAAAKA1qZAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAQi0lEQVR4nO3df6jd9X3H8eerMdXLWokuV4mJI3aEsFg2rRcJOEopZUltabI/ZBnrzB9CQCy0bEsx69jaPwpuYaUIU3BtMa5dJVCnQSqZpC39x9VdG22MmplOp/mBuaWEWsicpu/9cb5pj/Hm/og355zk83zAl+/3vM/3e77v80nufd3vj3tPqgpJUpveM+wGJEnDYwhIUsMMAUlqmCEgSQ0zBCSpYRcNu4HZLF26tFauXDnsNiTpvLF06VJ27969u6rWz7buyIfAypUrmZycHHYbknReSbJ0Lut5OkiSGmYISFLDDAFJapghIEkNMwQkqWEjf3eQJLXk4b2H2b77AEeOn+CqJWNsXbeajdcvP2f7MwQkaUQ8vPcw2x7ax4k3TwJw+PgJtj20D+CcBYGngyRpRGzffeDXAXDKiTdPsn33gXO2T0NAkkbEkeMn5lVfCIaAJI2Iq5aMzau+EAwBSRoRW9etZmzxorfVxhYvYuu61edsn14YlqQRcerir3cHSVKjNl6//Jx+0z+dp4MkqWGGgCQ1zBCQpIYZApLUMENAkhpmCEhSwwwBSWqYISBJDTMEJKlhhoAkNcwQkKSGGQKS1LA5h0CSRUn2Jnm0e3x5kseTvNjNL+tbd1uSg0kOJFnXV78hyb7uubuTZGHfjiRpPuZzJPBZ4Pm+x3cCe6pqFbCne0ySNcAm4FpgPXBPklN/IPteYAuwqpvWv6vuJUnvypxCIMkK4BPA1/rKG4Ad3fIOYGNf/cGqeqOqXgIOAjcmWQZcWlVPVFUBD/RtI0kagrkeCXwV+Dzwq77alVV1FKCbX9HVlwOv9q13qKst75ZPr79Dki1JJpNMTk1NzbFFSdJ8zRoCST4JHKuqp+b4mtOd568Z6u8sVt1XVRNVNTE+Pj7H3UqS5msunyx2E/CpJDcDlwCXJvkm8FqSZVV1tDvVc6xb/xBwdd/2K4AjXX3FNHVJ0pDMeiRQVduqakVVraR3wfd7VfVpYBewuVttM/BIt7wL2JTk4iTX0LsA/GR3yuj1JGu7u4Ju7dtGkjQE7+Yzhu8Cdia5DXgFuAWgqvYn2Qk8B7wF3FFVJ7ttbgfuB8aAx7pJkjQk6d2oM7omJiZqcnJy2G1I0nklyVNVNTHbev7GsCQ1zBCQpIYZApLUMENAkhpmCEhSwwwBSWqYISBJDTMEJKlhhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlqmCEgSQ0zBCSpYYaAJDXMEJCkhhkCktQwQ0CSGmYISFLDDAFJapghIEkNMwQkqWGGgCQ1zBCQpIYZApLUMENAkhpmCEhSwwwBSWqYISBJDTMEJKlhhoAkNcwQkKSGGQKS1LBZQyDJJUmeTPJMkv1JvtTVL0/yeJIXu/llfdtsS3IwyYEk6/rqNyTZ1z13d5Kcm7clSZqLuRwJvAF8tKr+ALgOWJ9kLXAnsKeqVgF7usckWQNsAq4F1gP3JFnUvda9wBZgVTetX7i3Ikmar1lDoHp+2T1c3E0FbAB2dPUdwMZueQPwYFW9UVUvAQeBG5MsAy6tqieqqoAH+raRJA3BnK4JJFmU5GngGPB4Vf0IuLKqjgJ08yu61ZcDr/ZtfqirLe+WT69Pt78tSSaTTE5NTc3j7UiS5mNOIVBVJ6vqOmAFvZ/qPzjD6tOd568Z6tPt776qmqiqifHx8bm0KEk6C/O6O6iqjgM/oHcu/7XuFA/d/Fi32iHg6r7NVgBHuvqKaeqSpCGZy91B40mWdMtjwMeAF4BdwOZutc3AI93yLmBTkouTXEPvAvCT3Smj15Os7e4KurVvG0nSEFw0h3WWATu6O3zeA+ysqkeTPAHsTHIb8ApwC0BV7U+yE3gOeAu4o6pOdq91O3A/MAY81k2SpCFJ70ad0TUxMVGTk5PDbkOSzitJnqqqidnW8zeGJalhhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlqmCEgSQ0zBCSpYYaAJDXMEJCkhhkCktQwQ0CSGmYISFLDDAFJapghIEkNMwQkqWGGgCQ1zBCQpIYZApLUMENAkhpmCEhSwwwBSWqYISBJDTMEJKlhhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlqmCEgSQ0zBCSpYYaAJDXMEJCkhhkCktSwWUMgydVJvp/k+ST7k3y2q1+e5PEkL3bzy/q22ZbkYJIDSdb11W9Isq977u4kOTdvS5I0F3M5EngL+Muq+j1gLXBHkjXAncCeqloF7Oke0z23CbgWWA/ck2RR91r3AluAVd20fgHfiyRpnmYNgao6WlU/7pZfB54HlgMbgB3dajuAjd3yBuDBqnqjql4CDgI3JlkGXFpVT1RVAQ/0bSNJGoJ5XRNIshK4HvgRcGVVHYVeUABXdKstB17t2+xQV1veLZ9en24/W5JMJpmcmpqaT4uSpHmYcwgkeR/wHeBzVfWLmVadplYz1N9ZrLqvqiaqamJ8fHyuLUqS5mlOIZBkMb0A+FZVPdSVX+tO8dDNj3X1Q8DVfZuvAI509RXT1CVJQzKXu4MCfB14vqq+0vfULmBzt7wZeKSvvinJxUmuoXcB+MnulNHrSdZ2r3lr3zaSpCG4aA7r3AT8ObAvydNd7a+Bu4CdSW4DXgFuAaiq/Ul2As/Ru7Pojqo62W13O3A/MAY81k2SpCFJ70ad0TUxMVGTk5PDbkOSzitJnqqqidnW8zeGJalhhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlqmCEgSQ0zBCSpYYaAJDXMEJCkhhkCktQwQ0CSGmYISFLDDAFJapghIEkNMwQkqWGGgCQ1zBCQpIYZApLUMENAkhpmCEhSwwwBSWqYISBJDbto2A1o+B7ee5jtuw9w5PgJrloyxtZ1q9l4/fJhtyVpAAyBxj289zDbHtrHiTdPAnD4+Am2PbQPwCCQGuDpoMZt333g1wFwyok3T7J994EhdSRpkAyBxh05fmJedUkXFkOgcVctGZtXXdKFxRBo3NZ1qxlbvOhttbHFi9i6bvWQOpI0SF4Ybtypi7/eHSS1yRAQG69f7jd9qVGeDpKkhhkCktQwQ0CSGjZrCCT5RpJjSZ7tq12e5PEkL3bzy/qe25bkYJIDSdb11W9Isq977u4kWfi3I0maj7kcCdwPrD+tdiewp6pWAXu6xyRZA2wCru22uSfJqfsP7wW2AKu66fTXlCQN2KwhUFU/BH5+WnkDsKNb3gFs7Ks/WFVvVNVLwEHgxiTLgEur6omqKuCBvm0kSUNyttcErqyqowDd/Iquvhx4tW+9Q11tebd8en1aSbYkmUwyOTU1dZYtSpJms9AXhqc7z18z1KdVVfdV1URVTYyPjy9Yc5KktzvbEHitO8VDNz/W1Q8BV/ettwI40tVXTFOXJA3R2YbALmBzt7wZeKSvvinJxUmuoXcB+MnulNHrSdZ2dwXd2reNJGlIZv2zEUm+DXwEWJrkEPB3wF3AziS3Aa8AtwBU1f4kO4HngLeAO6rq1B+rv53enUZjwGPdJEkaovRu1hldExMTNTk5Oew2JOm8kuSpqpqYbT1/Y1iSGmYISFLDDAFJapghIEkNMwQkqWGGgCQ1zBCQpIYZApLUMENAkhpmCEhSwwwBSWqYISBJDTMEJKlhhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlqmCEgSQ0zBCSpYYaAJDXMEJCkhhkCktQwQ0CSGmYISFLDDAFJapghIEkNMwQkqWGGgCQ1zBCQpIYZApLUMENAkhpmCEhSwy4adgPn2sN7D7N99wGOHD/BVUvG2LpuNRuvXz7stiRpJFyQIfA3D+/jm//xyjvqh4+fYNtD+wAMAkliCKeDkqxPciDJwSR3LvTrnykATjnx5km27z6w0LuVpPPSQEMgySLgn4CPA2uAP02yZiH3MVMAnHLk+ImF3KUknbcGfSRwI3Cwqv67qv4PeBDYMOAeuGrJ2KB3KUkjadAhsBx4te/xoa72Nkm2JJlMMjk1NbWgDQTYum71gr6mJJ2vBh0CmaZW7yhU3VdVE1U1MT4+vqAN/Nna3/GisCR1Bn130CHg6r7HK4AjC7mDSxaF/z35jlwB4Kt/cp0BIEl9Bn0k8J/AqiTXJHkvsAnYtZA7eOHLN3PJorcfcFyyKLx81ycMAEk6zUCPBKrqrSSfAXYDi4BvVNX+hd7PC1++eaFfUpIuSAP/ZbGq+i7w3UHvV5L0Tv7tIElqmCEgSQ0zBCSpYYaAJDUsVdPfUz8qkkwB/3MWmy4FfrbA7SyUUe4NRrs/ezt7o9yfvZ296fr7GUBVrZ9t45EPgbOVZLKqJobdx3RGuTcY7f7s7eyNcn/2dvbebX+eDpKkhhkCktSwCzkE7ht2AzMY5d5gtPuzt7M3yv3Z29l7V/1dsNcEJEmzu5CPBCRJszAEJKlhF2QInOsPsz+Lfl5Osi/J00kmu9rlSR5P8mI3v2xAvXwjybEkz/bVzthLkm3dOB5Ism5I/X0xyeFu/J5OcnPfcwPrL8nVSb6f5Pkk+5N8tqsPffxm6G3oY5fkkiRPJnmm6+1LXX3o4zZLf0Mfu779LUqyN8mj3eOFG7uquqAmen+i+qfAB4D3As8Aa4bc08vA0tNq/wDc2S3fCfz9gHr5MPAh4NnZegHWdON3MXBNN66LhtDfF4G/mmbdgfYHLAM+1C2/H/ivroehj98MvQ197Oh9ouD7uuXFwI+AtaMwbrP0N/Sx69vnXwD/CjzaPV6wsbsQjwRG4sPs52ADsKNb3gFsHMROq+qHwM/n2MsG4MGqeqOqXgIO0hvfQfd3JgPtr6qOVtWPu+XXgefpfUb20Mdvht7OZJC9VVX9snu4uJuKERi3Wfo7k4H2l2QF8Anga6f1sCBjdyGGwJw+zH7ACvj3JE8l2dLVrqyqo9D7AgauGFp3Z+5llMbyM0l+0p0uOnXoO7T+kqwErqf3U+NIjd9pvcEIjF13OuNp4BjweFWN1LidoT8YgbEDvgp8HvhVX23Bxu5CDIE5fZj9gN1UVR8CPg7ckeTDQ+5nrkZlLO8Ffhe4DjgK/GNXH0p/Sd4HfAf4XFX9YqZVp6md0/6m6W0kxq6qTlbVdfQ+V/zGJB+cYfWBj9sZ+hv62CX5JHCsqp6a6ybT1Gbs7UIMgXP+YfbzVVVHuvkx4N/oHZ69lmQZQDc/NrwOz9jLSIxlVb3WfZH+CvhnfnN4O/D+kiym9032W1X1UFceifGbrrdRGruun+PAD4D1jMi4nam/ERm7m4BPJXmZ3qntjyb5Jgs4dhdiCJzzD7OfjyS/leT9p5aBPwKe7Xra3K22GXhkOB3CDL3sAjYluTjJNcAq4MlBN3fqP3vnj+mN38D7SxLg68DzVfWVvqeGPn5n6m0Uxi7JeJIl3fIY8DHgBUZg3GbqbxTGrqq2VdWKqlpJ73vZ96rq0yzk2J3LK9rDmoCb6d0d8VPgC0Pu5QP0rtY/A+w/1Q/w28Ae4MVufvmA+vk2vUPbN+n91HDbTL0AX+jG8QDw8SH19y/APuAn3X/yZcPoD/hDeofWPwGe7qabR2H8Zuht6GMH/D6wt+vhWeBvZ/saGPC/65n6G/rYndbnR/jN3UELNnb+2QhJatiFeDpIkjRHhoAkNcwQkKSGGQKS1DBDQJIaZghIUsMMAUlq2P8DUjoUqVlZm0QAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "plt.scatter(x, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3687bd9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# what does it mean to have a different structure ?\n",
    "# if two proteins have similar structure, how different structures should they have?\n",
    "# can we label proteins to be used in supervised learning -> domains ?\n",
    "# If p1 and p2 are similar in structure, will they have same function?\n",
    "# mutations in the genome are harmful\n",
    "# hypothesis testing\n",
    "# significance doesn't imply biological significance\n",
    "# size effect -> how large n is \n",
    "\n",
    "\n",
    "# what is in the alphafold error file \n",
    "# sherlock \n",
    "# .gtf files -> code\n",
    "\n",
    "# permutation testing / bootstrap\n",
    "\n",
    "# paper that compare protein structures or functions\n",
    "# procrustes"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
